### PR TITLE
Fix: Adding tracing for the main functions

### PR DESCRIPTION
**Pull request by urgobalt**

### CODE REVIEW

**Review by barbariand — COMMENTED** [review]
Lgtm

File: parser/src/lib.rs [hunk 1]
```diff
@@ -24,33 +25,67 @@ pub mod token;
...
 pub async fn parse(text: &str) -> Result<Ast, Error> {
-    let (lexer_in, lexer_out): (Sender<Token>, Receiver<Token>) = mpsc::channel(32);
-    let (normalizer_in, normalizer_out): (Sender<Token>, Receiver<Token>) = mpsc::channel(32);
+    let span = trace_span!("parsing");
+    let _enter = span.enter();
+    debug!(text);
+    let channel = 32;
+    trace!("reading channel {channel:?}");
+    let (lexer_in, lexer_out): (Sender<Token>, Receiver<Token>) =
+        mpsc::channel(channel);
```
Comment (Alvinn8) [inline parser/src/lib.rs:34]: 32 is the capacity, so I think this variable name and the logging is a bit confusing. It sounds like 32 is some channel id, which it's not.

File: parser/src/lib.rs [hunk 2]
```diff
@@ -24,33 +25,67 @@ pub mod token;
...
+    debug!("successfully connected to lexer");
+    let (normalizer_in, normalizer_out): (Sender<Token>, Receiver<Token>) =
+        mpsc::channel(channel);
+    debug!("successfully connected to normalizer");
```
Comment (Alvinn8) [inline parser/src/lib.rs:38]: No connection happens here, the channels are just created. So printing that things were successfully connected seems a bit misleading.

File: parser/src/lib.rs [hunk 3]
```diff
@@ -24,33 +25,67 @@ pub mod token;
...
 
     let context = MathContext::new();
+    trace!("created MathContext");
     let lexer = Lexer::new(lexer_in);
+    trace!("created Lexer");
     let mut normalizer = Normalizer::new(lexer_out, normalizer_in);
+    trace!("created Normalizer");
     let mut parser = Parser::new(normalizer_out, context);
+    trace!("created Parser");
```
Comment (Alvinn8) [inline parser/src/lib.rs:47]: I'd maybe prefer if this tracing happened in the `new` functions, or not at all. Nothing can really go wrong here, we are just initializing structs.

File: parser/src/lib.rs [hunk 4]
```diff
@@ -24,33 +25,67 @@ pub mod token;
...
+    let lexer_handle =
+        tokio::spawn(async move { lexer.tokenize(&cloned_text).await });
+    trace!("spawning new normalize async task");
+    let normalizer_handle =
+        tokio::spawn(async move { normalizer.normalize().await });
+    trace!("spawning new parser async task");
     let parser_handle = tokio::spawn(async move { parser.parse().await });
     match lexer_handle.await {
         Err(e) => {
+            error!("lexer task failed");
```
Comment (Alvinn8) [inline parser/src/lib.rs:60]: Maybe print the error? (In other places too)

File: parser/src/lib.rs [hunk 5]
```diff
@@ -24,33 +25,67 @@ pub mod token;
...
+                Ok(()) => {
+                    debug!("normalizer task finished");
+                    match parser_handle.await {
+                        Err(e) => {
+                            error!("parser task failed");
+                            Err(Error::Parser(e))
+                        }
+                        Ok(ast) => {
+                            debug!("parser task finished");
+                            error!("ast generation failed");
```
Comment (Alvinn8) [inline parser/src/lib.rs:82]: Nothing failed here? It returns a result, which might be Err or Ok.

Comment (Alvinn8) [inline parser/src/lib.rs:82]: Nothing failed here? `ast` is a Result, which might be Err or Ok.

**Review by Alvinn8 — CHANGES_REQUESTED** [review]
Nice, this will help with debugging. See my comments though.

**Pull request closed without merging**